generators/python-v2/ast/src/PythonFile.ts

```diff
@@ -44,47 +48,99 @@ export class PythonFile extends AstNode {
      * Helper Methods
      *******************************/
 
-    private writeComments(writer: Writer): void {
-        this.comments.forEach((comment) => {
-            comment.write(writer);
+    private updateWriterRefNameOverrides({
+        writer,
+        uniqueReferences
+    }: {
+        writer: Writer;
+        uniqueReferences: Map<string, { modulePath: ModulePath; references: Reference[]; referenceNames: Set<string> }>;
+    }): void {
+        const reservedNames = this.getReservedNames();
+        const references: Reference[] = Array.from(uniqueReferences.values()).flatMap(({ references }) => references);
+
+        // Build up a map of refs to their name overrides, keeping track of howmany times we've seen a name as we go.
+        const completeRefPathsToNameOverrides: Record<string, { name: string; isAlias: boolean }> = {};
+        const nameUsageCounts: Record<string, number> = {};
+
+        // Initialize counts for reserved names
+        reservedNames.forEach((name) => {
+            nameUsageCounts[name] = 1;
         });
 
-        if (this.comments.length > 0) {
-            writer.newLine();
-        }
+        references.forEach((reference) => {
+            const name = reference.alias ?? reference.name;
+            const refIdentifier = reference.getCompletePath();
+
+            // Get current count for this name, accounting for reserved names used by statements in the file
+            const currentCount = nameUsageCounts[name] ?? 0;
+            nameUsageCounts[name] = currentCount + 1;
+
+            // For reserved names or names we've seen before, use an auto-generated alias with a naming convention
+            // of a numbered suffix.
+            if (reservedNames.has(name) || currentCount > 0) {
+                completeRefPathsToNameOverrides[refIdentifier] = {
+                    name: `${name}_${currentCount}`,
+                    isAlias: true
+                };
+            } else {
+                completeRefPathsToNameOverrides[refIdentifier] = {
+                    name,
+                    isAlias: !!reference.alias
+                };
+            }
+        });
+
+        writer.setRefNameOverrides(completeRefPathsToNameOverrides);
     }
 
-    private getImportName(reference: Reference): string {
-        const name = reference.name;
-        const alias = reference.alias;
-        return `${name}${alias ? ` as ${alias}` : ""}`;
+    private getReservedNames(): Set<string> {
+        const reservedNames = new Set<string>();
+
+        this.statements.forEach((statement) => {
+            if (statement instanceof Class) {
+                reservedNames.add(statement.name);
+            } else if (statement instanceof Method) {
+                reservedNames.add(statement.name);
+            } else if (statement instanceof Field) {
+                reservedNames.add(statement.name);
+            }
+        });
+
+        return reservedNames;
     }
```

Comment:
I'd update the name here - when I think of reserved names I think of keywords like `def`, `__init__`, etc. Maybe just `getUsedNames`?

Reply:
Yep good call 👍